# Viral Genome Assembly from Long Reads

The aims of this repository is to reproduce the results generated by a study by [Montecillo et al.](https://journals.asm.org/doi/10.1128/mra.00719-22) on viral genome assembly using minION reads.

### Workflow

1. Retrieval of sequencing reads (PRJNA857442) and swine reference genome (GCF_000003025.6)
2. Quality control with `porechop` (v0.2.4)
3. Host read removal (decontamination) with `minimap2` (v2.17-r941)
4. _de novo_ assembly with Flye (v2.4.2)
5. Refinement of the draft assembly with `medaka` (v1.4.4)
6. Whole genome alignment with `mafft` (v7.490)
7. ML tree inference with `iqtree` (v2.2.0)


### Environment Setup

Create a yaml file containing a list of dependencies. Include the versions to ensure reproducibility.

```yaml
name: asfv-assembly
channels:
  - bioconda
  - conda-forge
dependencies:
  - flye
  - iqtree=2.2
  - mafft
  - medaka
  - minimap2=2.17
  - porechop=0.2.4
  - python=3.11
```

Create environment using `micromamba` and activate:
```bash
micromamba create -f env.yml
```

### Assembly Pipeline

All steps for the assembly pipeline are organized as bash functions under the `scripts` directory. A short description of each bash script is provided below:

- `globals.sh`: contains all the global variables used in the pipeline
- `00_run_complete_pipeline.sh`: an executable for running the complete assembly pipeline
- `01_download_data.sh`: downloads the sequencing reads from the SRA and the reference genome from NCBI
- `02_trim_reads.sh`: detects and trims adapters from long reads
- `03_map_and_filter.sh`: aligns the trimmed reads to the reference genome and filter for host read contaminants
- `04_assemble.sh`: perform _de novo_ assembly on the unmapped reads converted into FASTQ format
